general/kdtree.hpp

```diff
@@ -0,0 +1,765 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_KDTREE_HPP
+#define MFEM_KDTREE_HPP
+
+#include <vector>
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <cmath>
+
+#include "../linalg/vector.hpp"
+
+namespace mfem
+{
+
+namespace KDTreeNorms
+{
+
+/// Evaluates l1 norm of a vector.
+template <typename Tfloat, int ndim>
+struct Norm_l1
+{
+   Tfloat tm;
```

Comment:
I will move tm in the function - nice point!

I didn't consider complex numbers. For complex numbers, or quaternions and perhaps octonions, it might be more beneficial to split them in the KDTree.